_zh-cn/overviews/scala3-book/domain-modeling-fp.md

```diff
@@ -0,0 +1,477 @@
+---
+title: 函数式领域建模
+type: section
+description: This chapter provides an introduction to FP domain modeling with Scala 3.
+num: 22
+previous-page: domain-modeling-oop
+next-page: methods-intro
+---
+
+
+本章介绍了在 Scala 3 中使用函数式编程 (FP) 进行领域建模。
+当使用 FP 对我们周围的世界进行建模时，您通常会使用以下 Scala 构造：
+
+- 枚举
+- 样例类
+- Traits
+
+> 如果您不熟悉代数数据类型 (ADT) 及其泛型版本 (GADT)，您可能需要先阅读 [代数数据类型][adts] 部分，然后再阅读本节。
+
+## 介绍
+
+在 FP 中，*数据* 和*对该数据的操作* 是两个独立的东西；您不必像使用 OOP 那样将它们封装在一起。
+
+这个概念类似于数值代数。
+当您考虑值大于或等于零的整数时，您有一*组*可能的值，如下所示：
+
+````
+0, 1, 2 ... Int.MaxValue
+````
+
+忽略整数的除法，对这些值可能的*操作*是：
+
+````
++, -, *
+````
+
+FP设计以类似的方式实现：
+
+- 你描述你的值的集合（你的数据）
+- 您描述了对这些值起作用的操作（您的函数）
+
+> 正如我们将看到的，这种风格的程序推理与面向对象的编程完全不同。
+> FP 中的数据只**是**：
+> 将功能与数据分离，让您无需担心行为即可检查数据。
+
+在本章中，我们将为披萨店中的“披萨”建模数据和操作。
+您将看到如何实现 Scala/FP 模型的“数据”部分，然后您将看到几种不同的方式来组织对该数据的操作。
+
+## 数据建模
+
+在 Scala 中，描述编程问题的数据模型很简单：
+
+- 如果您想使用不同的替代方案对数据进行建模，请使用 枚举
+- 如果您只想对事物进行分组（或需要更细粒度的控制），请使用 样例类
+
+### 描述替代方案
+
+简单地由不同的选择组成的数据，如面饼大小、面饼类型和馅料，使用 Scala 3 枚举进行简洁的建模：
+
+```scala
+enum CrustSize:
+  case Small, Medium, Large
+
+enum CrustType:
+  case Thin, Thick, Regular
+
+enum Topping:
+  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
+```
+
+> 描述不同选择的数据类型（如 `CrustSize`）有时也称为_归纳类型_。
```

Comment:
Maybe we should use sum types, I am not sure.

Reply:
keep it for now.